macos/resource_fork.ksy

```diff
@@ -0,0 +1,593 @@
+
+      In practice,
+      the data blocks in newly created resource files are usually contiguous.
+      When existing resources are shortened,
+      the Mac OS resource manager leaves unused space where the now removed resource data was,
+      as this is quicker than moving the following resource data into the newly freed space.
+      Such unused space may be cleaned up later when the resource manager "compacts" the resource file,
+      which happens when resources are removed entirely,
+      or when resources are added or grown so that more space is needed in the data area.
+  resource_map:
+    pos: header.ofs_resource_map
+    type: resource_map
+    size: header.len_resource_map
+    doc: The resource file's resource map.
+types:
+  file_header:
+    seq:
+      - id: ofs_data_blocks
+        type: u4
+        doc: |
+          Offset of the resource data area,
+          from the start of the resource file.
+
+          In practice,
+          this should always be `256`,
+          i. e. the resource data area should directly follow the application-specific data area.
+      - id: ofs_resource_map
+        type: u4
+        doc: |
+          Offset of the resource map,
+          from the start of the resource file.
+
+          In practice,
+          this should always be `ofs_data_blocks + len_data_blocks`,
+          i. e. the resource map should directly follow the resource data area.
+      - id: len_data_blocks
+        type: u4
+        doc: |
+          Length of the resource data area.
+      - id: len_resource_map
+        type: u4
+        doc: |
+          Length of the resource map.
+
+          In practice,
+          this should always be `_root._io.size - ofs_resource_map`,
+          i. e. the resource map should extend to the end of the resource file.
+    doc: |
+      Resource file header,
+      containing the offsets and lengths of the resource data area and resource map.
+  data_block:
+    seq:
+      - id: len_data
+        type: u4
+        doc: |
+          The length of the resource data stored in this block.
+      - id: data
+        size: len_data
+        doc: |
+          The data stored in this block.
+    doc: |
+      A resource data block,
+      as stored in the resource data area.
+
+      Each data block stores the data contained in a resource,
+      along with its length.
+  resource_map:
+    seq:
+      - id: reserved_file_header_copy
+        type: file_header
+        doc: Reserved space for a copy of the resource file header.
+      - id: reserved_next_resource_map_handle
+        type: u4
+        doc: Reserved space for a handle to the next loaded resource map in memory.
+      - id: reserved_file_reference_number
+        type: u2
+        doc: Reserved space for the resource file's file reference number.
+      - id: file_attributes
+        type: file_attributes
+        size: sizeof<file_attributes> # Force creation of a substream
```

Comment:
Normally I wouldn't have used an explicit `size` here at all. I only needed to add one so that a new substream is created for this field, so that the `as_int` instance in `file_attributes` works correctly (hence the comment next to it). Otherwise I want the same behavior as if the field had no explicit `size` at all (i. e. let the type read all the data that it needs and continue reading directly after that), so I tried to simulate that using `sizeof<...>`.

I understand what you mean with checking that the fields in the structure add up to the expected size, but I don't think setting a manually hardcoded `size` really helps with that. No matter if you have a manual `size` or not, if you get the number/size of fields in the type wrong, you'll get errors when parsing data. The manual `size` might make the error messages a little more useful if you accidentally make the type too long, but it will silence errors if the type is too short. There's also an extra chance of errors if you miscalculate the manual `size`.

I don't see other specs in KSF do this either - for example `archive/gzip` doesn't have `size`s on its flag integers:

https://github.com/kaitai-io/kaitai_struct_formats/blob/db23b8c0fb214d7e0259409b6fe3bc76bd020dfc/archive/gzip.ksy#L33-L46

It might not be a bad idea to check a type's calculated size against an expected byte count (especially for things like bit flag integer fields, where the size is explicitly given in the documentation), but `size` isn't the right tool for that. Ideally, checks like this would be done at the type level and statically by the compiler, not when the type is used and at runtime. This would of course require new syntax and support by the compiler and is not something we can do in specs right now.

Reply:
> There's also an extra chance of errors if you miscalculate the manual `size`.

The classical approach (i.e. used by at least 95% of manually written parsers I assume) is to read a n-byte integer, which represents the packed bitfield, and then using bitwise operators to extract the individual fields. The specifications usually count on this as well: first, you'll get the information that there's a field `flags` that has 2 bytes in sizes, and _then_ you'll see the internal structure of the `flags` bitfield. I haven't seen a format specification that would just show a list of bit integers in `flags` with just the number of bits the individual fields occupy, without any byte size, saying _dear programmers, summing the bit sizes up to get the `flags` size in bytes is left as an exercise for the reader_.

That said, you usually get the byte size in advance, so the _miscalculation_ IMHO doesn't occur often, if you don't invent your own format and you don't know yet how many bits you'll need in the end.

> The manual `size` might make the error messages a little more useful if you accidentally make the type too long, but it will silence errors if the type is too short.

Of course, but `size: sizeof<..>` does the same. When the bits have a total of 15 bits, `sizeof<..>` will be 2. But there are cases where this behavior is intended. As I say above, when you're developing a `.ksy` spec for some format, you may postpone dissecting the internal structure of the `flags` to the point when you'll need it, so you first just mark they occupy 2 bytes and skip it. Or imagine that you're trying to figure out the structure of a bitfield during reverse-engineering. You can save yourself the trouble of ensuring that the fields have a total size of 16 bits by constantly updating the bit size of the "unknown purpose" field at the end by just wrapping the type with flags in a substream with fixed byte size.

Reply:
> Normally I wouldn't have used an explicit `size` here at all. I only needed to add one so that a new substream is created for this field, so that the `as_int` instance in `file_attributes` works correctly (hence the comment next to it). Otherwise I want the same behavior as if the field had no explicit `size` at all (i. e. let the type read all the data that it needs and continue reading directly after that), so I tried to simulate that using `sizeof<...>`.

I can't recommend using the current substream in a subtype which ends with a bit-sized integer field, because if the subtype `seq` accidentally doesn't end on a byte boundary (8, 16, 24, ...), it can wreak havoc sometimes. This is due to the bug relating to the byte alignment after bit-sized integer followed by a byte aligned field. The current implementation treats all subtypes as byte-aligned and doesn't look what's inside them. So if a subtype does not end on a byte boundary and it's followed by a byte-aligned field (so the `alignToByte()` method should be called, but doesn't happen of course), it will eventually cause weird things first time you parse a bit-sized integer in the same stream (see https://github.com/kaitai-io/kaitai_struct/issues/155#issuecomment-616219006).

Putting the subtype in a substream will ensure that everything after it will be "aligned" automatically, because it doesn't affect the `bits` and `bitsLeft` properties of the parent stream.